File: packages/python/src/alumnium/server/agents/changes_analyzer_agent.py
from langchain_core.language_models import BaseChatModel

from ..logutils import get_logger
from .base_agent import BaseAgent

logger = get_logger(__name__)


class ChangesAnalyzerAgent(BaseAgent):
    def __init__(self, llm: BaseChatModel):
        super().__init__()
        self.llm = llm

    def invoke(self, diff: str) -> str:
        logger.info("Starting changes analysis:")
        indented_diff = "\n".join(f"  {line}" for line in diff.splitlines())
        logger.debug(f"  -> Diff:\n{indented_diff}")

        message = self._invoke_chain(  # type: ignore[assignment]
            self.llm,
            [
                ("system", self.prompts["system"]),
                ("human", self.prompts["user"].format(diff=diff)),
            ],
        )

        content = message.text.replace("\n\n", " ")  # type: ignore[reportAttributeAccessIssue]
        logger.info(f"  <- Result: {content}")
        logger.info(f"  <- Usage: {message.usage_metadata}")  # type: ignore[reportAttributeAccessIssue]

        return content

File: packages/core/src/server/agents/ChangesAnalyzerAgent.ts
import { BaseChatModel } from "@langchain/core/language_models/chat_models";
import z from "zod";
import { pythonicFormat } from "../../pythonic/pythonicFormat.js";
import { getLogger } from "../../utils/logger.js";
import type { LlmContext } from "../LlmContext.js";
import { BaseAgent } from "./BaseAgent.js";

const logger = getLogger(import.meta.url);

export namespace ChangesAnalyzerAgent {
  export type Meta = z.infer<typeof ChangesAnalyzerAgent.Meta>;
}

export class ChangesAnalyzerAgent extends BaseAgent {
  static Meta = z.object({
    kind: z.literal("changes-analyzer"),
  });

  static readonly EXCLUDE_ATTRIBUTES = new Set(["id"]);
  llm: BaseChatModel;

  constructor(llmContext: LlmContext, llm: BaseChatModel) {
    super(llmContext);
    this.llm = llm;
  }

  async invoke(diff: string): Promise<string> {
    logger.info("Starting changes analysis:");
    logger.debug(this.formatLog("in", "Diff"), { detail: diff });

    const meta: ChangesAnalyzerAgent.Meta = {
      kind: "changes-analyzer",
    };

    const response = await this.invokeChain(
      this.llm,
      [
        ["system", this.prompts.system],
        ["human", pythonicFormat(this.prompts.user, { diff })],
      ],
      meta,
    );

    const content = response.content.replaceAll("\n\n", " ");

    this.logData(logger, "out", {
      Result: content,
      Usage: response.usage,
    });

    return content;
  }
}
